py/runai_model_streamer/runai_model_streamer/distributed_streamer/distributed_streamer.py

```diff
@@ -0,0 +1,533 @@
+    FileStreamer,
+    FileChunks,
+)
+
+from runai_model_streamer.distributed_streamer.partition import (
+    partition,
+    get_total_number_of_chunks,
+    get_partition_policy,
+    log_partition_info,
+)
+
+from runai_model_streamer.s3_utils.s3_utils import (
+    S3Credentials,
+    is_s3_path,
+    is_gs_path,
+)
+
+from timeit import default_timer as timer
+
+DEFAULT_BROADCAST_TIMEOUT = timedelta(seconds=600)
+
+class DistributedStreamer:
+    def __init__(self) -> None:
+        self.file_streamer = FileStreamer()
+        self.distributed_streamer = _distributedStreamer(self.file_streamer)
+        self.is_distributed = False
+        self.num_processes_on_node = None
+        self.group_size = 1
+        self.my_global_rank = 0
+        self.groups_by_ranks = []
+        self.broadcast_timeout = DEFAULT_BROADCAST_TIMEOUT
+        self.max_chunk = 0
+
+    def __enter__(self) -> DistributedStreamer:
+        self.file_streamer.__enter__()
+        self.distributed_streamer.__enter__()
+        return self
+
+    def __exit__(self, exc_type: any, exc_value: any, traceback: any) -> None:
+        try:
+            self.distributed_streamer.__exit__(exc_type, exc_value, traceback)
+        finally:
+            # Ensure the second __exit__ is always called
+            self.file_streamer.__exit__(exc_type, exc_value, traceback)
+        return
+    
+    def get_group_size(self) -> int:
+        if not dist.is_initialized():
+            return 1
+        return dist.get_world_size()
+
+    def get_cuda_free_memory(self) -> int:
+        if not torch.cuda.is_available():
+            return 0
+        free_memory, total_memory = torch.cuda.mem_get_info()
+        return free_memory
+
+    def set_is_distributed(self, is_distributed: bool, path: str, device: str) -> None:
+        # check if distributed streaming should be used
+
+        if not is_distributed:
+            self.is_distributed = False
+            return
+
+        # environment variable to override default distributed streaming
+        # by default, use distributed streaming only for object storage paths
+        enable_dist = os.environ.get("RUNAI_STREAMER_DIST")
+        if enable_dist == "0":
+            self.is_distributed = False
+        elif enable_dist == "1":
+            self.is_distributed = True
+        elif enable_dist == "auto" or enable_dist is None:
+            if path is not None:
+                self.is_distributed = is_s3_path(path) or is_gs_path(path)
+            else:
+                self.is_distributed = False
+        else:
+            raise ValueError(f"Invalid value for RUNAI_STREAMER_DIST: {enable_dist}")
+
+        # check if torch distributed is initialized and there are more than one process
+        if self.is_distributed:
+            self.is_distributed = dist.is_initialized()
+            if not self.is_distributed:
+                print(f"Note: torch distributed is not initialized")
```

Comment:
From the user perspective it should be distributed, but the fact that torch dist is not initialized is what makes us fall back to not distributed.

I think we ought to add the text "we fall back to not distributed", because the user wont be able to understand that only by the text "torch distributed is not initialized"